tests/frameworks/module.py

```diff
@@ -0,0 +1,223 @@
+import sys
+import typing as t
+from importlib.abc import Loader
+from importlib.machinery import ModuleSpec
+from importlib.util import find_spec
+from pathlib import Path
+from types import ModuleType
+from typing import Any, Callable, Dict, Optional, Set, Union, cast
+
+
+def origin(module):
+    # type: (ModuleType) -> str
+    """Get the origin source file of the module."""
+    try:
+        assert module.__file__ is not None
+        orig = str(Path(module.__file__).resolve())  # type: ignore[type-var]
+    except (AttributeError, TypeError):
+        # Module is probably only partially initialised, so we look at its
+        # spec instead
+        try:
+            orig = str(Path(module.__spec__.origin).resolve())  # type: ignore
+        except (AttributeError, ValueError, TypeError):
+            orig = None
+
+    if orig is not None and Path(orig).is_file():
+        if orig.endswith(".pyc"):
+            orig = orig[:-1]
+        return orig
+
+    return "<unknown origin>"
+
+
+def find_loader(fullname):
+    # type: (str) -> Optional[Loader]
+    return getattr(find_spec(fullname), "loader", None)
+
+
+class _ImportHookChainedLoader(Loader):
```

Comment:
This is based on existing code that was already tested and known to work. Perhaps the better strategy would be to implement `get_code` to intercept all module code objects and recurse over them